### PR TITLE
feat: support ranges in Lucene filters and field groups

### PR DESCRIPTION
**Pull request by kaya-david**
## Description

* add support for inclusive numeric Lucene range expressions using square brackets, such as age:[18 TO 65].
* support integer and floating-point ranges for both regular search fields and field group syntax.
* support negative range boundaries by normalizing Luqum's Prohibit nodes.
* reject exclusive ranges, non-numeric or non-finite boundaries, and reversed ranges with descriptive errors.
* add parametrized tests covering valid ranges, boundary values, invalid expressions, and both supported syntax variants.

## Assignee
- [x] The changes adhere to the [contribution guidelines](https://github.com/fkie-cad/Logprep/blob/main/CONTRIBUTING.md)
- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings (e.g. flake8/mypy/pytest/...) other than deprecations

### Documentation
- [x] [README.md](https://github.com/fkie-cad/Logprep/blob/main/README.md) is up-to-date
- [x] [CHANGELOG.md](https://github.com/fkie-cad/Logprep/blob/main/CHANGELOG.md) is up-to-date
- [x] [Documentation](https://github.com/fkie-cad/Logprep/tree/main/doc/source) (doc/source) is up-to-date
- [x] [Preview for docs](#readthedocs-preview) looks fine
- [x] [Notebooks](https://github.com/fkie-cad/Logprep/tree/main/doc/source/development/notebooks) are up-to-date and functional
- [x] [Examples](https://github.com/fkie-cad/Logprep/tree/main/examples) are up-to-date and functional (including compose & k8s)

### Code Quality
- [x] Patch test coverage > 95% and does not decrease
- [x] New code uses correct & specific type hints

### How did you verify that the changes work in practice?

* pytest

## Reviewer
- [ ] The code is readable/maintainable and follows the [contribution guidelines](https://github.com/fkie-cad/Logprep/blob/main/CONTRIBUTING.md)
- [ ] [Documentation](#documentation) is up-to-date
- [ ] Tests (unit & acceptance) are meaningful and not over-mocked

---
<a name="readthedocs-preview"></a>The rendered docs for this PR can be found [here](https://logprep--976.org.readthedocs.build/).

### CODE REVIEW

File: logprep/filter/lucene_filter.py [hunk 1]
```diff
@@ -58,6 +58,38 @@
...
+..  code-block:: yaml
+    :linenos:
+    :caption: Example
+
+    filter: 'temperature:([18.5 TO 25.0])'
+
+
+Only closed numeric ranges using square brackets are currently supported.
+Exclusive ranges using curly brackets, open boundaries using :code:`*`, and
+non-numeric boundaries are not supported.
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py]: Do we want to allow square brackets? I think that should be a relativly easy implementation

File: logprep/filter/lucene_filter.py [hunk 2]
```diff
@@ -275,6 +313,14 @@ def _parse_tree(self, tree: luqum.tree) -> FilterExpression:
             # pylint: enable=no-value-for-parameter
         if isinstance(tree, Group):
             return self._parse_tree(tree.children[0])
+
+        if isinstance(tree, Range):
+            if self._last_search_field is None:
+                raise LuceneFilterError(f'The expression "{str(tree)}" is invalid!')
+
+            key = get_dotted_field_list(self._last_search_field)
+            return self._parse_range(key, tree)
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py:390]: Is this actually ever run? In my testing I think this couldnt be reached / wasnt, but im not sure

File: logprep/filter/lucene_filter.py [hunk 3]
```diff
@@ -288,6 +334,97 @@ def _parse_tree(self, tree: luqum.tree) -> FilterExpression:
...
 
+    def _parse_range(self, key: Sequence[str], expr: Range) -> FilterExpression:
+        expression = str(expr)
+        lower_token, upper_token = expr.children
+
+        if not expression.startswith("[") or not expression.endswith("]"):
+            raise LuceneFilterError(
+                f'The expression "{expression}" is invalid. '
+                "Only inclusive ranges using square brackets are supported."
+            )
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py]: This seems, overengineered, does expression not expose this in a nicer way?

File: logprep/filter/lucene_filter.py [hunk 4]
```diff
@@ -288,6 +334,97 @@ def _parse_tree(self, tree: luqum.tree) -> FilterExpression:
...
+                key,
+                lower_bound,
+                upper_bound,
+            )
+
+        try:
+            lower_bound_float = float(lower_value)
+            upper_bound_float = float(upper_value)
+        except ValueError as error:
+            raise LuceneFilterError(f'The expression "{expression}" is invalid!') from error
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py]: We do want string ranges, so timestamps of ISO 8601 Format, are possible

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,9 @@
 ### Breaking
 
 ### Features
+* add support for inclusive numeric Lucene range filters using square brackets.
+* add support for integer and floating-point ranges in regular fields and field groups.
+* add validation for reversed, non-numeric, non-finite, and unsupported exclusive ranges.
```
Comment (Pablu23) [inline CHANGELOG.md]: I would summarize this all as one point if possible

**Review by Pablu23 — COMMENTED** [review]
Looks good so far, had a few questions and notes

File: logprep/filter/lucene_filter.py (hunk 2 above)
Comment (kaya-david) [reply logprep/filter/lucene_filter.py:390]: It does. Setting a breakpoint in this code and running the tests shows it.

File: logprep/filter/lucene_filter.py (hunk 1 above)
Comment (kaya-david) [reply logprep/filter/lucene_filter.py]: Implemented square brackets

File: logprep/filter/lucene_filter.py [hunk 6]
```diff
@@ -288,6 +402,211 @@ def _parse_tree(self, tree: luqum.tree) -> FilterExpression:
...
+        upper_value: str,
+        include_lower_bound: bool,
+        include_upper_bound: bool,
+        expr: Range,
+    ) -> FloatRangeFilterExpression | None:
+        """Create a float range expression if both boundaries are finite floats."""
+        if LuceneTransformer._is_integer_range_boundary(
+            lower_value
+        ) or LuceneTransformer._is_integer_range_boundary(upper_value):
+            return None
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py]: What is this for? Through ordering alone this shoulndt be possible, and If I call this with an int explicitly I might want it to behave like a float. also this doesnt really make a difference for floats to ints?

File: logprep/filter/lucene_filter.py [hunk 7]
```diff
@@ -288,6 +402,211 @@ def _parse_tree(self, tree: luqum.tree) -> FilterExpression:
...
+        key: Sequence[str],
+        lower_value: str,
+        upper_value: str,
+        include_lower_bound: bool,
+        include_upper_bound: bool,
+        expr: Range,
+    ) -> StringRangeFilterExpression:
+        """Create a lexicographic string range expression for non-numeric boundaries."""
+        if lower_value == "*" or upper_value == "*":
+            raise LuceneFilterError(f'The expression "{expr}" is invalid!')
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py:496]: What is this check for?

File: logprep/filter/lucene_filter.py [hunk 8]
```diff
@@ -288,6 +402,211 @@ def _parse_tree(self, tree: luqum.tree) -> FilterExpression:
...
+
+        if LuceneTransformer._is_numeric_range_boundary(
+            lower_value
+        ) or LuceneTransformer._is_numeric_range_boundary(upper_value):
+            raise LuceneFilterError(f'The expression "{expr}" is invalid!')
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py]: Same as above, I would remove this as ordering will disallow it anyway and calling it explicitly should not raise an error, as it is possible

File: logprep/filter/lucene_filter.py [hunk 9]
```diff
@@ -288,6 +402,211 @@ def _parse_tree(self, tree: luqum.tree) -> FilterExpression:
...
+        """Return whether a range boundary can be interpreted as a finite float."""
+        if LuceneTransformer._is_integer_range_boundary(value):
+            return False
+
+        try:
+            numeric_value = float(value)
+        except ValueError:
+            return False
+
+        return math.isfinite(numeric_value)
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py]: This is unnecessary then

File: CHANGELOG.md [hunk 10]
```diff
@@ -2,6 +2,8 @@
 ### Breaking
 
 ### Features
+* add support for validated integer, floating-point, and lexicographic string Lucene range filters in regular fields and field groups.
+* add support for quoted string range boundaries, allowing values with Lucene special characters such as ISO-8601 timestamps with timezone offsets.
```
Comment (Pablu23) [inline CHANGELOG.md]: I would put them both into one summary of what happened, "add lucene range support for integer, float and strings, supporting ISO-8601 timestamps"

File: logprep/filter/lucene_filter.py [hunk 11]
```diff
@@ -288,6 +402,211 @@ def _parse_tree(self, tree: luqum.tree) -> FilterExpression:
...
+                upper_value,
+                include_lower_bound,
+                include_upper_bound,
+                expr,
+            )
+
+            if range_filter_expression is not None:
+                return range_filter_expression
+
+        raise LuceneFilterError(f'The expression "{expression}" is invalid!')
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py]:
```suggestion
    def _parse_range(self, key: Sequence[str], expr: Range) -> FilterExpression:
        lower_value = self._get_range_boundary_value(expr.low)
        upper_value = self._get_range_boundary_value(expr.high)

        for range_parser in (
            self._parse_integer_range,
            self._parse_float_range,
            self._parse_string_range,
        ):
            range_filter_expression = range_parser(
                key,
                lower_value,
                upper_value,
                expr.include_low,
                expr.include_high,
                expr,
            )

            if range_filter_expression is not None:
                return range_filter_expression

        raise LuceneFilterError(f'The expression "{expr}" is invalid!')
```

**Review by Pablu23 — COMMENTED** [review]
Thanks for the work, looks good so far, there are a few points I would like to see a change or I dont agree with but thats it

File: logprep/filter/lucene_filter.py (hunk 7 above)
Comment (kaya-david) [reply logprep/filter/lucene_filter.py:496]: This check explicitly rejects open-ended Lucene ranges such as `[* TO 10]` or `[10 TO *]`.

The current implementation only supports closed ranges with concrete lower and upper boundaries. Without this check, `*` could fall through to the string range parser and be treated as a normal lexicographic string boundary, which would make open ranges appear to be supported accidentally.

File: logprep/filter/lucene_filter.py (hunk 8 above)
Comment (kaya-david) [reply logprep/filter/lucene_filter.py]: This check prevents numeric-looking boundaries from falling through to the string range parser.

Without it, expressions such as `[1 TO bar]` or `[0 TO 10.5]` could be interpreted as lexicographic string ranges after the integer and float parsers declined them. That would make mixed or ambiguous range boundaries appear to be supported accidentally.

The intention here is that unquoted numeric-looking boundaries are handled only by the numeric range parsers. String ranges should only use non-numeric boundaries, or quoted boundaries when the value contains Lucene special characters.

File: logprep/filter/lucene_filter.py (hunk 6 above)
Comment (kaya-david) [reply logprep/filter/lucene_filter.py]: That make sense, removed it.

File: logprep/filter/lucene_filter.py (hunk 11 above)
Comment (kaya-david) [reply logprep/filter/lucene_filter.py]: fixed

File: logprep/filter/lucene_filter.py (hunk 9 above)
Comment (kaya-david) [reply logprep/filter/lucene_filter.py]: Fixed/changed

File: CHANGELOG.md (hunk 10 above)
Comment (kaya-david) [reply CHANGELOG.md]: Done

File: logprep/filter/lucene_filter.py (hunk 7 above)
Comment (Pablu23) [reply logprep/filter/lucene_filter.py:496]: I see, then im fine with this, do we have it in the docs that we dont allow open-ended ranges for now?

File: logprep/filter/lucene_filter.py (hunk 8 above)
Comment (Pablu23) [reply logprep/filter/lucene_filter.py]: Okay than that seems good to me aswell, thanks for the clarification

File: logprep/filter/lucene_filter.py [hunk 12]
```diff
@@ -288,6 +402,171 @@ def _parse_tree(self, tree: luqum.tree) -> FilterExpression:
...
+        expr: Range,
+    ) -> StringRangeFilterExpression | None:
+        """Create a lexicographic string range expression for non-numeric boundaries."""
+        if lower_value == "*" or upper_value == "*":
+            raise LuceneFilterError(f'The expression "{expr}" is invalid!')
+
+        if LuceneTransformer._is_finite_numeric_range_boundary(
+            lower_value
+        ) or LuceneTransformer._is_finite_numeric_range_boundary(upper_value):
+            return None
```
Comment (Pablu23) [inline logprep/filter/lucene_filter.py:501]: Maybe raise here instead of returning None, and then waiting for it to raise in the upper method, to keep consistent with the other methods

Comment (kaya-david) [reply logprep/filter/lucene_filter.py:501]: I would keep returning `None` here.

The range parsers are tried in order: integer, float, string. Returning `None` means “this parser is not responsible for this boundary type”, while raising is reserved for cases that are invalid for the parser that did accept the type.

Numeric-looking boundaries are not valid string range boundaries, but they are also not string-parser-specific errors. Returning `None` lets `_parse_range` produce the final generic invalid-range error if no parser can handle the expression. This keeps the parser chain consistent with `_parse_integer_range` and `_parse_float_range`, which also return `None` when the boundary type does not match.

Comment (Pablu23) [reply logprep/filter/lucene_filter.py:501]: Yeah true that is fair

**Review by Pablu23 — APPROVED** [review]
LGTM thank you for the work
